### PR TITLE
Simplify default generated module

### LINKED ISSUE #5: Add option for creating a basic module
The current output is pretty complex for a first-time user, can we add an option to make a simpler "Hello World" style module?

### PR DESCRIPTION
**Pull request by digitalinfinity**
Change the default generated module to be a simpler module.
The newly generated module now looks like a hello world example.

Added a new --intermediate flag to generate a more complicated but realistic module

Fixes #5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,12 +16,24 @@ npm install -g yo
 npm install -g generator-napi-module
 ```
 
+Create a directory for your N-API module:
+```bash
+mkdir <moduleName>
```
Comment (kfarnung) [inline README.md:21]: You also need to cd into the directory, right?

Comment (digitalinfinity) [reply README.md:21]: Bah, fair - will update

**Review by kfarnung — APPROVED** [review]
LGTM

**Comment by digitalinfinity** [comment]
Thanks! Now let's see how this CI works 😝

**Comment by kfarnung** [comment]
Hmm:

Jest did not exit one second after the test run has completed.
This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.

You may want to cancel build #5:  https://travis-ci.org/digitalinfinity/generator-napi-module/builds/440927245

It looks like it’s hung somewhere and will have to wait 10 minutes of no console output to be cancelled by Travis.

-Kyle

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10
[HxS - 17763 - 16.0.11012.31451]

________________________________

**Comment by digitalinfinity** [comment]
Odd. Jest passes on my machine and running it with --detechOpenHandles doesn't really behave differently:
```
C:\code\napi\generator-napi-module>npm run test

> generator-napi-module@0.2.1 pretest C:\code\napi\generator-napi-module
> tslint --project .


> generator-napi-module@0.2.1 test C:\code\napi\generator-napi-module
> jest --detectOpenHandles

 PASS  __tests__/app.js (6.578s)
  generator-napi-module:app
    √ creates files (2ms)
  generator-napi-module:app:ts
    √ creates files
  generator-napi-module:app:inter
    √ creates files (1ms)
  generator-napi-module:app:inter:ts
    √ creates files

Test Suites: 1 passed, 1 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        6.674s, estimated 10s
Ran all test suites.
```

**Comment by digitalinfinity** [comment]
Hmm, they all actually failed with the same error - in the new test (--intermediate + --typescript), it looks like lib/binding.ts isn't there- i guess this must be an xplat issue?

**Comment by digitalinfinity** [comment]
No! It's a logic issue- now I don't understand how the test passes...

**Comment by kfarnung** [comment]
Maybe some stale files?  I'm not sure how the tests execute specifically, but I'm able to repro the failure here.

**Comment by digitalinfinity** [comment]
Yeah- I assumed the tests would clean up after every test runs but now I'm not so sure

**Comment by kfarnung** [comment]
But hey, the CI is paying off already :)

File: src/app/index.ts [hunk 2]
```diff
@@ -179,21 +188,36 @@ module.exports = class extends Generator {
...
+    let bindingWrapperSrc = (genIntermediate ?
+                                "lib/binding_intermediate.js" :
+                                bindingWrapperTarget);
     if (genTypeScript) {
-      bindingWrapper = "lib/binding.ts";
+      if (genIntermediate) {
+        bindingWrapperSrc = "lib/binding_intermediate.ts";
+      }
+
+      bindingWrapperTarget = "lib/binding.ts";
```
Comment (kfarnung) [inline src/app/index.ts:230]: Aren't we forgetting to get lib/binding.ts into the Src as well?

Comment (digitalinfinity) [reply src/app/index.ts:230]: yes- bah, what are these tests doing...I had manually tested the combinations before to confirm that the modules did the right thing and built but then simplified the logic and relied on the unit tests to do the validation- now i'm concerned :(

Comment (kfarnung) [reply src/app/index.ts:230]: Yeah, that's definitely a problem.  Here's my proposed fix for that logic: https://github.com/kfarnung/generator-napi-module/commit/03db6c4752669e90636ecc15f4e11c705b1fdc69

Comment (digitalinfinity) [reply src/app/index.ts:230]: oh- i see- js is ts 😛 ... it was simply copying lib/binding.js to lib/binding.ts- and so the file with the right name was there, just the wrong contents. I still dont know why the first set of tests passed locally but failed on CI but hopefully this is right now. AFAICT, it's not a cleanup issue- every test run gets a new SHA-based directory name, so one test's execution should not affect the next from a file perspective

Comment (kfarnung) [reply src/app/index.ts:230]: You're making sure to build before you test, right?

**Comment by kfarnung** [comment]
No idea what's going on with the node/6 and node/8 legs of the CI, they were passing before.

File: src/app/index.ts (hunk 2 above)
Comment (digitalinfinity) [reply src/app/index.ts:230]: yup

**Comment by digitalinfinity** [comment]
Yeah, very puzzling

**Comment by jschlight** [comment]
I am attempting an `npm pack` on a clone of the `simple_module` branch in order to install it for testing on my development system. There are errors in the console log that appear related to `typescript` that I don't understand how to address:

```
vecto:generator-napi-module jim$ npm pack

> generator-napi-module@0.2.1 prepare /Users/jim/N-API-Repositories/generator-napi-module
> npm run build


> generator-napi-module@0.2.1 build /Users/jim/N-API-Repositories/generator-napi-module
> tsc

error TS2688: Cannot find type definition file for 'node'.

error TS2688: Cannot find type definition file for 'yeoman-generator'.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! generator-napi-module@0.2.1 build: `tsc`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the generator-napi-module@0.2.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jim/.npm/_logs/2018-11-02T21_06_12_887Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! generator-napi-module@0.2.1 prepare: `npm run build`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the generator-napi-module@0.2.1 prepare script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jim/.npm/_logs/2018-11-02T21_06_12_908Z-debug.log
```

I do have Yeoman installed globally.

**Comment by jschlight** [comment]
Doh! Never mind. Operator error. I'm continuing to test...

**Comment by jschlight** [comment]
Okay. I believe I've got the current `simple_module` branch installed. I'm seeing errors when running `yo napi-module `:

```
vecto:generator-test jim$ node -v
v10.8.0
vecto:generator-test jim$ npm -v
6.2.0
vecto:generator-test jim$ npm list -g --depth 0
/Users/jim/.nvm/versions/node/v10.8.0/lib
├── generator-napi-module@0.2.1
├── npm@6.2.0
└── yo@2.0.5

vecto:generator-test jim$ yo napi-module
internal/modules/cjs/loader.js:583
    throw err;
    ^

Error: Cannot find module './clone.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
    at Function.Module._load (internal/modules/cjs/loader.js:507:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at Object.<anonymous> (/Users/jim/.nvm/versions/node/v10.8.0/lib/node_modules/yo/node_modules/graceful-fs/graceful-fs.js:4:13)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
vecto:generator-test jim$ 
```

**Comment by gabrielschulhof** [comment]
@digitalinfinity @kfarnung
@jschlight, @NickNaso and I tried generating from this branch on Node.js 10.10 and the generator did not ask whether we wanted the simpler version of the package or the more complex version, but went ahead and generated the simpler version.

**Comment by gabrielschulhof** [comment]
As it stands, adding `--intermediate` to the generation will actually generate the complex version, and leaving it out will generate the simple version. Thus, perhaps we should rename `--intermediate` to `--complex` and use the intermediate by default.

**Comment by gabrielschulhof** [comment]
@digitalinfinity @kfarnung would it be possible to have these changes by the end of your workday today (Monday)? We would base the flow of the workshop on whether this PR has landed or not.

**Comment by kfarnung** [comment]
Sorry, what's the request here?  Just renaming "--intermediate" to "--complex"?

**Comment by kfarnung** [comment]
I pushed a change to add a menu:

```
? Choose a template (Use arrow keys)
❯ Hello World
  Object Wrap
```

**Comment by NickNaso** [comment]
Thanks @kfarnung it seems that all checks have been passed now. @digitalinfinity  it could be great have the generator published in time for our workshop.

**Comment by jschlight** [comment]
I am updating the conference tutorials to reflect Kyle's latest changes. It would great if this PR could land and the module be published sometime Tuesday in order for the conference participants to benefit from your work during the session on Wednesday.

**Comment by kfarnung** [comment]
I published a new version on NPM: https://www.npmjs.com/package/generator-napi-module

**Comment by jschlight** [comment]
Awesome, Kyle. Thank you!

I've published the tutorial we plan to use for tomorrow's workshop [here](https://napi.inspiredware.com/getting-started/first.html). I'm open to any constructive criticism.

**Pull request merged**